scylla/src/transport/connection_keeper.rs

```diff
@@ -113,60 +128,118 @@ impl ConnectionKeeper {
 
     pub async fn use_keyspace(
         &self,
-        keyspace_name: &VerifiedKeyspaceName,
+        keyspace_name: VerifiedKeyspaceName,
     ) -> Result<(), QueryError> {
-        // ConnectionKeeper doesn't have reconnecting yet so this will be ok for now
-        // TODO: Modify once ConnectionKeeper gets reconnecting
+        let (response_sender, response_receiver) = tokio::sync::oneshot::channel();
 
-        self.get_connection()
-            .await?
-            .use_keyspace(keyspace_name)
+        self.use_keyspace_channel
+            .send(UseKeyspaceRequest {
+                keyspace_name,
+                response_chan: response_sender,
+            })
             .await
+            .expect("Bug in ConnectionKeeper::use_keyspace sending");
+        // Other end of this channel is in the Worker, can't be dropped while we have &self to _worker_handle
+
+        response_receiver.await.unwrap() // ConnectionKeeperWorker always responds
     }
 }
 
 impl ConnectionKeeperWorker {
-    pub async fn work(self) {
-        let connect_result = self.open_new_connection().await;
-
-        match connect_result {
-            Ok((conn, _error_receiver)) => {
-                let _ = self
-                    .conn_state_sender
-                    .send(ConnectionState::Connected(conn.clone()));
-
-                let new_shard_info: Option<ShardInfo> = conn.get_shard_info().clone();
-
-                if let Some(sender) = &self.shard_info_sender {
-                    // Ignore sending error
-                    // If no one wants to get shard_info that's OK
-                    // If lock is poisoned do nothing
-                    if let Ok(sender_locked) = sender.lock() {
-                        let _ = sender_locked.send(new_shard_info);
-                    }
-                }
-            }
-            Err(e) => {
-                let _ = self.conn_state_sender.send(ConnectionState::Broken(e));
-            } // TODO: Wait for connection to fail, then create new, loop it
-        };
+    pub async fn work(mut self) {
+        // Reconnect at most every 8 seconds
+        let reconnect_cooldown = tokio::time::Duration::from_secs(8);
+        let mut last_reconnect_time;
+
+        loop {
+            last_reconnect_time = tokio::time::Instant::now();
+
+            // Connect and wait for error
+            let current_error: QueryError = self.run_connection().await;
+
+            // Mark the connection as broken, wait cooldown and reconnect
+            let _ = self
+                .conn_state_sender
+                .send(ConnectionState::Broken(current_error));
+
```

Comment:
If the `ConnectionKeeperWorker` notices that either the keyspace channel or error receiver returns `None`, it should mean that the whole connection is meant to be closed, no? Now, I think that if `ConnectionKeeperWorker` is quick enough, it will make a short attempt to reconnect because of the retry logic in loop of `ConnectionKeeperWorker::work`.

Unless I'm missing something, we should explicitly cause the loop to stop right here if we detect a situation I described.

Reply:
It would stop soon because its `RemoteHandle` would get dropped and on first async `poll` worker would get dropped too before even getting a chance to reconnect I think.

I added shutdown logic to make this explicit.

Reply:
Maybe you are right, I don't know if it's possible and how likely in practice to happen, but keep in mind that somebody can run our driver using multithreaded version of the `tokio` runtime. `ConnectionKeeper` and `ConnectionKeeperWorker` are running on different tasks, so they can be scheduled to run on different CPU cores, in separate threads. You can't assume that events happening between preemption points will be observed as if they happened atomically - here, you can't assume this about keyspace channel being closed and `RemoteHandle` being dropped.

It's better to be explicit and not rely on timing issues.